Subscribe to logical replication

jgaskins opened the pull request:
The basic idea is that you create a replication subscription with `PG.connect_replication`. It's similar to `PG.connect_listen` but, unlike a `LISTEN` connection where every message is the same type (it just yields strings), each message can be a different type. So instead of passing a block you pass a `PG::Replication::Handler` instance with `received` methods defined for receiving different types of replication messages:

```crystal
class MyHandler
  include PG::Replication::Handler

  # Override this to tell the Postgres server where you are in the WAL.
  # Make sure you `yield` in between any setup and teardown.
  def received(data : PG::Replication::XLogData, connection : PG::Replication::Connection)
    yield
    connection.last_wal_byte_applied = data.wal_end
    connection.last_wal_byte_flushed = data.wal_end
  end

  # New relation defined
  def received(relation : PG::Replication::Relation)
  end

  # Record updated
  def received(update : PG::Replication::Update)
  end
end

subscriber = PG.connect url,
  handler: MyHandler.new,
  publication_name: "my_publication",
  slot_name: "my_replication_slot"
```

The default is to do nothing when receiving a `WALMessage`.

Most of the work here happens in `XLogData` events, which contain the `WALMessage`s that people will actually care about. Everything else seems to be protocol meta. The objects passed to the handler's `received` methods are the `XLogData`'s `message` value and are decoded based on message formats specified in [the Postgres docs](https://www.postgresql.org/docs/current/protocol-logicalrep-message-formats.html).

There are still some things I need and/or want to do here:
- [x] Implement message types
  - [x] Begin
  - [x] Message
  - [x] Commit
  - [x] Origin
  - [x] Relation
  - [x] Type
  - [x] Insert
  - [x] Update
  - [x] Delete
  - [x] Truncate
- [ ] Allow decoding `TupleData` as other types (rather than raw `Bytes`)
- [ ] Allow starting the WAL from somewhere other than the beginning
- [x] Implement keepalive responses
- [x] Since logical replication requires a change to `postgresql.conf`, the specs need to be opt-in or enabled only when logical replication is enabled

The initial spec does successfully invoke the handler's methods for the `CREATE TABLE`, `INSERT`, and `UPDATE` commands.

jgaskins committed:
Add logical replication handling

jgaskins commented on src/pg/replication.cr:
```diff
@@ -0,0 +1,353 @@
+module PG::Replication
```
I can't tell yet if this should be `LogicalReplication` or if physical replication uses the same concepts.

jgaskins commented on src/pg/replication.cr:
```diff
@@ -0,0 +1,353 @@
+module PG::Replication
+  module Handler
+    def received(frame)
+    end
```
This is needed so that you don't get compile-time errors when you don't define all possible messages. It doesn't have to be untyped, though. It might be better to be explicit that this is for `WALMessage` objects only.

jgaskins commented on src/pg/replication.cr:
```diff
@@ -0,0 +1,353 @@
+      end
+    end
+
+    def received(frame : CopyBoth)
+    end
```
I don't know what to do here or if there's anything _to_ do. This is the first message that comes in after you send the `START_REPLICATION` command.

jgaskins commented on src/pg/replication.cr:
```diff
@@ -0,0 +1,353 @@
+    def received(frame : CopyBoth)
+    end
+
+    def received(frame : CopyData)
+      received frame.data
+    end
```
This is the message that contains the `XLogData`

jgaskins commented on src/pg/replication.cr:
```diff
@@ -0,0 +1,353 @@
+      received frame.data
+    end
+
+    def received(data : XLogData)
+      handler.received data.message
+    end
```
… and the `XLogData` contains the `WALMessage`.

jgaskins commented on src/pg/replication.cr:
```diff
@@ -0,0 +1,353 @@
+    end
+
+    protected def read_string(io : IO) : String
+      io.read_line('\0', chomp: true)
```
This usage makes me think `read_line` is a misnomer for this method. It's reading up to a delimiter.

This may also be the least efficient part of this whole PR. Reading a string up to a delimiter may involve internal reallocations for longer strings as the buffer keeps filling up.

jgaskins commented on src/pg/replication.cr:
```diff
@@ -0,0 +1,353 @@
+        when 'n'
+          nil
+        when 'u'
+          UnchangedTOASTValue.new
```
I haven't tested this yet, but the docs say that [an unchanged TOASTed value is not sent](https://www.postgresql.org/docs/current/protocol-logicalrep-message-formats.html#PROTOCOL-LOGICALREP-MESSAGE-FORMATS-TUPLEDATA), so I'm assuming it's just a marker to indicate that the previous value is unchanged to avoid sending huge strings over the wire unnecessarily.

jgaskins commented on src/pq/connection.cr:
```diff
@@ -188,45 +189,44 @@ module PQ
     end
 
+    def start_replication_frame_loop(publication_name : String, slot_name : String, &block : PG::Replication::Frame ->)
+      command = "START_REPLICATION SLOT #{slot_name} LOGICAL 0/0 (proto_version '1', binary 'true', publication_names '#{publication_name}')"
```
A few notes on this line:

- There are currently [4 `proto_version` values](https://www.postgresql.org/docs/current/protocol-logical-replication.html#PROTOCOL-LOGICAL-REPLICATION-PARAMS), but anything other than `1` requires Postgres 14+ and changes how some `WALMessage`s are decoded. Supporting these dynamically would be more complicated, so I'm just supporting the baseline out of the gate.
- This is currently hardcoded to start at the beginning of the WAL (`0/0`). The `PG.listen_replication` method should almost certainly be updated to allow passing in other LSNs to start from a different point in the WAL history.

will replied:
FWIW postgres 13 is EOL in 3 months I think. Still starting support with only 1 is probably fine.

jgaskins replied:
Nice. Main addition of in v2 is transaction ids on various actions, which could be useful for some that might want this functionality. I wouldn't say it's necessary since the events are linear (all CRUD actions map to the transaction id from the most recent `Begin` WAL message), but it might still be useful.

jgaskins commented on src/pq/connection.cr:
```diff
@@ -188,45 +189,44 @@ module PQ
-        handle_error frame
-        true
-      elsif frame.is_a?(Frame::NotificationResponse)
-        handle_notification frame
-        true
-      elsif frame.is_a?(Frame::NoticeResponse)
-        handle_notice frame
-        true
-      elsif frame.is_a?(Frame::ParameterStatus)
-        handle_parameter frame
-        true
-      else
-        false
-      end
+      false
     end
 
-    private def handle_error(error_frame : Frame::ErrorResponse)
+    private def handle_async_frames(error_frame : Frame::ErrorResponse)
       expect_frame Frame::ReadyForQuery if @established
       notice_handler.call(error_frame.as_notice)
       raise PQError.new(error_frame.fields)
     end
 
-    private def handle_notice(frame : Frame::NoticeResponse)
-      notice_handler.call(frame.as_notice)
+    private def handle_async_frames(frame : Frame::NotificationResponse)
+      notification_handler.call(frame.as_notification)
+      true
     end
 
-    private def handle_notification(frame : Frame::NotificationResponse)
-      notification_handler.call(frame.as_notification)
+    private def handle_async_frames(frame : Frame::NoticeResponse)
+      notice_handler.call(frame.as_notice)
+      true
     end
 
-    private def handle_parameter(frame : Frame::ParameterStatus)
+    private def handle_async_frames(frame : Frame::ParameterStatus)
```
The reason I updated this was because I was originally going to try to reuse `read_async_frame_loop`. That didn't work, but along the way I realized that the `if frame.is_a?` was basically a `case` statement, and even that was equivalent to just using an overloaded method. When I backed out my attempt, I just left it because I think this refactor still simplifies the code a bit.

jgaskins commented on src/pq/connection.cr:
```diff
@@ -239,18 +239,21 @@ module PQ
       ]
+      if replication
+        startup_args << "replication" << replication
+      end
```
`["replication", "true"]` enables physical-replication mode and `["replication", "database"]` enables logical-replication mode ([docs](https://www.postgresql.org/docs/current/protocol-replication.html)). `PG::Replication::Connection` [currently hardcodes it to `"database"`](https://github.com/will/crystal-pg/pull/301/files#diff-9db8c24b86773bc96353706cc4aec1b44c47b8ee0e06c8a2828d818b9605e004R19) to enable logical replication.

jgaskins reviewed (commented):
Just some notes about the current state of things. Some of them are subject to change.

will commented:
I've not done much with logical replication myself, so I cant comment on the details of the protocol here, but it all seems super cool!

jgaskins commented:
I've had to add Debezium to our stack to pull some data into our data warehouse. All the trouble I've had getting it working got me wondering how easy or difficult it would be to just consume the WAL ourselves and emit a similar output structure. _Completely_ nerd-sniped.

> I cant comment on the details of the protocol here

I'm hoping nobody will need to be all that concerned with it. 😄 It's not *super* complicated, but I want to come up with a good set of tests that illustrate that replication is working/broken by just performing actions and asserting on things appearing in the stream of logical replication events.

jgaskins committed:
Make the handler update last bytes flushed/applied

jgaskins commented:
Alrighty, I think this is at least ready for initial review. Feel free to let me know if anything is unclear.

will commented:
Awesome, thanks! I think this is all good to go. Could you add a changelog entry too please first?

jgaskins commented:
Sweet, will do. I did find a bug the other night, though, that I still need to address. Specifically: error frames aren't being decoded on this connection. While looking into that, I realized the error frame is [this error](https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-ERRORRESPONSE), which you've already got here. Same message structure, which makes sense.

Then I wondered, since that's the same as `ErrorResponse`, if the [`CopyBoth` message](https://github.com/will/crystal-pg/pull/301/files#diff-9db8c24b86773bc96353706cc4aec1b44c47b8ee0e06c8a2828d818b9605e004R158-R177) is documented on that page, too. And [it is](https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-COPYBOTHRESPONSE). That's when I realized I _could've_ reused the same connection structure as `LISTEN` like I originally intended.

I think, for the moment, I'll just reimplement the error frame here (the pattern I used in implementing these messages turned out pretty decent) and refactor toward reuse later.

jgaskins committed:
Extract replication messages to individual files

As a single file, it was much longer than necessary.

jgaskins committed:
Handle disconnections from the server

If we get too far behind on processing data from the server, it could
disconnect us. This commit handles that disconnection by reconnecting
from and picking up where we left off.

jgaskins committed:
Make keepalive backstop interval configurable

It was hardcoded to 10 seconds before because that is quite comfortably
lower than the Postgres default `wal_sender_timeout` of 60 seconds, but
some Postgres installations configure that setting to lower than that.
For example, EDB's `cloudnative-pg` sets it to 5 seconds. If we fall
behind on the WAL stream and don't process the `KeepAlive` message
from the server within that time, Postgres will disconnect us.

Making this configurable allows apps to receive some setting to set it
to whatever is needed. For example, with `cloudnative-pg`, you'd set it
to 2-3 seconds.
